src/celeritas/global/alongstep/AlongStep.hh

```diff
@@ -0,0 +1,134 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2022 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/global/alongstep/AlongStep.hh
+//! \brief Along-step function and helper classes
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "corecel/Assert.hh"
+#include "corecel/Macros.hh"
+#include "corecel/OpaqueId.hh"
+#include "corecel/Types.hh"
+#include "orange/Types.hh"
+#include "celeritas/Types.hh"
+
+namespace celeritas
+{
+//---------------------------------------------------------------------------//
+//! Within-kernel state for step limits and action
+struct AlongStepLocalState
+{
+    //! Final step limit (true path length) and action
+    StepLimit step_limit;
+    //! Smooth (straight/curved) step movement without accounting for MSC
+    real_type geo_step{};
+};
+
+//---------------------------------------------------------------------------//
+/*!
+ * Perform the along-step action using helper functions.
+ *
+ * \tparam MH MSC helper, e.g. \c detail::NoMsc
+ * \tparam MP Propagator factory, e.g. \c detail::LinearPropagatorFactory
+ * \tparam AE Energy loss applier, e.g. \c detail::NoElossApplier
+ */
+template<class MH, class MP, class AE>
+inline CELER_FUNCTION void along_step(MH&&                 msc,
+                                      MP&&                 make_propagator,
+                                      AE&&                 apply_eloss,
+                                      CoreTrackView const& track)
+{
+    // TODO: scope the 'views' so that the lifetimes don't overlap between this
+    // function and helper class functions
+    auto sim = track.make_sim_view();
+
+    // True step is the actual path length traveled by the particle, including
+    // within-step MSC
+    AlongStepLocalState local;
+    local.step_limit = sim.step_limit();
+    CELER_ASSERT(local.step_limit);
+    if (local.step_limit.step == 0)
```

Comment:
This is also exactly what we're doing now, so I'll add a checkbox on #459 to revisit this.

Reply:
We'll need to remove the assertion that `step > 0` since we _can_ still have zero step length here. As @whokion said, Geant4 invokes at-rest processes at the beginning of the step, while we currently do it at the end, along with the rest of the discrete actions. If a positron loses all energy over the step, it _might_ annihilate at the end of the step, or the discrete action might be rejected in the integral method (since the process is sampled based on pre-step macro xs, so e.g. it might sample bremsstrahlung and then reject the interaction). If that happens, the positron will take one more step and the step length will be zero at this point. I think this might also be why we see a difference in the number of steps per track between Celeritas and Geant4 for positrons:
![steps_per_track_testem3_zoom](https://user-images.githubusercontent.com/6426426/180472032-22e39d7a-4b97-4760-811c-d6dd1f4620cc.png)

Reply:
Aha, good point @amandalund .